### PR TITLE
Send notification on join-organisation request

### LINKED ISSUE #411: Send notification on join organization request
We need to send a notification when a "join organization" request is issued. This notification should go out to all current org members.

### PR DESCRIPTION
**Pull request by aquibbaig**
Fixes #411

### CODE REVIEW

File: site/src/Controller/OrganizationController.php [hunk 1]
```diff
@@ -215,6 +220,23 @@ public function joinAction($organizationName, OrganizationRepository $repository
         $member->setUser($user);
         $member->setPermission(OrganizationMember::PERMISSION_REQUEST);
         $em = $this->getDoctrine()->getManager();
+        // Find Organisation members
+        $orgMembers = $repository->findOneByName($organizationName)->getMemberUsers();
```
Comment (imphil) [inline site/src/Controller/OrganizationController.php]: This should be limited to org admins, normal users cannot accept/reject a request to join.

File: site/src/Controller/OrganizationController.php [hunk 2]
```diff
@@ -215,6 +220,23 @@ public function joinAction($organizationName, OrganizationRepository $repository
...
+        // Queue notifications to each member
+        $notification = new Notification();
+        foreach ($orgMembers as $recipient) {
+            // Prevent mailing the currentUser
+            if ($recipient !== $this->getUser()) {
```
Comment (imphil) [inline site/src/Controller/OrganizationController.php]: Wouldn't this mean that the user requesting to join an org is already part of it? I think this check can be removed.

File: site/src/Controller/OrganizationController.php [hunk 3]
```diff
@@ -215,6 +220,23 @@ public function joinAction($organizationName, OrganizationRepository $repository
...
+                $notification->setType("organisation_join")
+                    ->setSubject("Join organisation request")
+                    ->setRecipient($recipient)
+                    ->setMessage(
+                        $member->getUser()->getUsername()." has requested to join your organisation.
+                    Kindly go to the Librecores website to accept/reject the request."
```
Comment (imphil) [inline site/src/Controller/OrganizationController.php]: Please include the organization name, and a link to the org admin page (or directly link to the accept/reject actions). And please fix the indentation.

**Comment by aquibbaig** [comment]
Thanks, @imphil for the review. I will look into the following changes.

**Comment by aquibbaig** [comment]
@imphil Does these settings make sense? I believe the message can be improved further. I will test them out at once.

File: site/src/Controller/OrganizationController.php [hunk 4]
```diff
@@ -215,6 +220,20 @@ public function joinAction($organizationName, OrganizationRepository $repository
...
+        // Find Organisation members
+        $orgAdmin = $repository->findOneByName($organizationName)->getCreator();
+        // Queue notifications to the organisation admin
+        $notification = new Notification();
+        $notification->setType("organisation_join_request")
+            ->setSubject("Organisation join request")
+            ->setRecipient($orgAdmin)
+            ->setMessage(
+                $member->getUser()->getUsername()." has requested to join ".
+                $organizationName . ". Kindly go to https://www.librecores.org/"
```
Comment (imphil) [inline site/src/Controller/OrganizationController.php]: Please use the router component to construct this URL, and ensure that the domain is not hardcoded but depends on the configuration (or in this case, on the request).

File: site/src/Controller/OrganizationController.php [hunk 5]
```diff
@@ -215,6 +220,20 @@ public function joinAction($organizationName, OrganizationRepository $repository
         $member->setUser($user);
         $member->setPermission(OrganizationMember::PERMISSION_REQUEST);
         $em = $this->getDoctrine()->getManager();
+        // Find Organisation members
+        $orgAdmin = $repository->findOneByName($organizationName)->getCreator();
```
Comment (imphil) [inline site/src/Controller/OrganizationController.php]: Please send this notification to all admins (can be more than one). They are OrganizationMember entities with the PERMISSION_ADMIN permission set.

**Comment by aquibbaig** [comment]
Thanks @imphil on the follow up. I almost had forgot about that `url` parameter we created during last summer and you're completely right about using that. I have updated the PR. WIll test it out soon

File: site/src/Controller/OrganizationController.php (hunk 4 above)
Comment (aquibbaig) [reply site/src/Controller/OrganizationController.php]: +1

**Pull request merged**